Vulkan async pipeline creation

hrydgard opened the pull request:
An experiment trying to reduce shader stutters by doing Vulkan pipeline creation on a separate thread. Later could be done on multiple threads to further utilize many-cores. (Warning: Pipelines using the same vs/ps should be on the same thread, sequentially!)

Ideally, compilation to SPIR-V and shader module creation should also be moved to this thread. Additionally, texture upscaling and so on could also be done on a thread.

hrydgard committed:
VulkanRenderManager: Add deferred pipeline creation (to get it off the CPU thread)

hrydgard committed:
Vulkan: Add a single background thread for pipeline creation

Add proper waits for compile-done

Don't rely on non-standard struct initialization of classes

Attempt to drain the pipeline compile queue before destroying the PipelineManager.

Vulkan: Bump the cache version for testing

unknownbrackets commented on GPU/Vulkan/PipelineManagerVulkan.cpp:
```diff
@@ -31,8 +31,11 @@ void PipelineManagerVulkan::Clear() {
-			vulkan_->Delete().QueueDeletePipeline(value->pipeline);
+		if (value->pipeline) {
+			VkPipeline pipeline = value->pipeline->pipeline;
+			vulkan_->Delete().QueueDeletePipeline(pipeline);
```
Hm, I wonder if we need to wait if it is not set yet, in case we try to delete it while compiling or something?

-[Unknown]

hrydgard replied:
Yeah, I somehow didn't think of that case, needs checks indeed.

hrydgard commented on ext/native/thin3d/VulkanQueueRunner.cpp:
```diff
@@ -778,11 +786,45 @@ void VulkanQueueRunner::PerformRenderPass(const VKRStep &step, VkCommandBuffer c
+		case VKRRenderCommand::BIND_GRAPHICS_PIPELINE:
+		{
+			VKRGraphicsPipeline *pipeline = c.graphics_pipeline.pipeline;
+			if (!pipeline->pipeline) {
```
I wonder if there's any possibility of a race here - if it gets done and notifies *just* before we lock and do the wait below - but on the other hand, if it's done, it will never enter the while loop anyway. Hm. It probably is safe...

unknownbrackets replied:
Yeah, I think that sounds safe too.

-[Unknown]

hrydgard commented:
Rebased on master, added draining of the compile queue before shutting down the PipelineCache 
which should fix the proposed issue with pipeline deletions.

However, while this should reduce shader stutters, I don't see that big of a difference. Smaller stutters may be entirely eliminated while larger one where compiles take several frames are only slightly reduced. Might still be worth merging to later add the ability to use more cores to compile shaders - but another avenue might be to simply reduce the number of unique shaders. Uniformly dynamic branching is no longer very expensive, I think it can be assumed to be cheap on Vulkan-capable chips.

hrydgard commented:
Rebased on master, still working fine. I'll merge after 1.10 is out.

unknownbrackets commented:
Does this interact well with lost/restore?  Do we need to drain the queue on lost too?

-[Unknown]

hrydgard commented:
Right, maybe should look into that, too..

hrydgard commented:
I'm postponing this, too risky for 1.12.

Gamemulatorer commented:
How about this now?

unknownbrackets commented:
On OpenGL, could we in theory create a separate GL context on another thread, compile shaders there, and download their binaries for the main thread to use?  Wonder if that would be worth the trouble.

-[Unknown]

hrydgard commented:
Highly doubt it would be worth the trouble. We should instead probably be working towards trying to reduce the number of unique shaders, checking uniform constants is pretty cheap and on modern hardware nearly free. Unfortunately it's kinda awkward on the really old mobile GPUs..

Gamemulatorer commented:
This is not gonna include to https://github.com/hrydgard/ppsspp/milestone/34?

Because when I click the v1.3.0 milestone it get me to future-prio

hrydgard commented:
@Gamemulatorer I haven't really begun sorting issues from future-prio into 1.3.0 yet

unknownbrackets commented on Common/GPU/Vulkan/VulkanQueueRunner.cpp:
```diff
@@ -841,9 +841,11 @@ void VulkanQueueRunner::LogRenderPass(const VKRStep &pass, bool verbose) {
 				INFO_LOG(G3D, "  (Removed)");
 				break;
-
-			case VKRRenderCommand::BIND_PIPELINE:
```
Should we keep this logging since we kept BIND_PIPELINE?

-[Unknown]

hrydgard replied:
That was probably a merge mistake, I'll take a look.

unknownbrackets commented on Common/GPU/Vulkan/VulkanRenderManager.cpp:
```diff
@@ -22,6 +22,47 @@
+	if (result == VK_INCOMPLETE) {
+		// Bad return value seen on Adreno in Burnout :(  Try to ignore?
+		// Create a placeholder to avoid creating over and over if something is broken.
+		pipeline = VK_NULL_HANDLE;
```
Hm, it might be nice to log the shader if we can?  I guess it's farther away now.

-[Unknown]

hrydgard replied:
Added a comment for now..

hrydgard commented:
I plan to get this in real soon as a first step, but there's more work to do. I looked at this in Superluminal at the classic major hitch on the first punch in Tekken 6, where the change in lighting causes a lot of shader compiles. It's about a 30% improvement, which isn't bad, but it's wasting a lot of potential parallelism by having a single shader compile thread. Making better use of the new thread pool/task system might be possible.

Additionally, I found that it's not just pipeline creation that costs, even generating the SPIR-V from GLSL takes 1+ms per shader, which adds up when there are many simultaneous new shaders, and that's on a fast desktop CPU (although this part is quite a bit less costly than the pipeline creation).

Another separate improvement might be to make the vertex part more uber-shader-like (branching in the shader instead of generating separate ones) to simply reduce the shader count...

hrydgard committed:
Add a hidden setting to turn off the shader cache, for shader compile performance work

hrydgard committed:
Comment improvement, restore accidental deletion

hrydgard committed:
Minor refactor

Gamemulatorer commented:
This should belong https://github.com/hrydgard/ppsspp/milestone/34?closed=1 1.13.0 Milestone.

unknownbrackets commented:
Thanks.

-[Unknown]
